Remove repeat decompose and map to ttnn's repeat ops respectively

ashokkumarkannan1 opened the pull request:
Fix https://github.com/tenstorrent/tt-forge-fe/issues/683

### Root cause
The mismatch of number of elements between Input and output tensors is due to broadcast op which is converted to tm op in `pre lowering pass` which is part of this `convert_broadcast_ops_to_tms`. On further investigation, found that repeat op is decomposed to broadcast in `forge_passes.py` and since [repeat](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api/ttnn.repeat.html#ttnn-repeat) & [repeat_interleave](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api/ttnn.repeat_interleave.html#ttnn.repeat_interleave) op present in ttnn we can directly map to it through MLIR.

Since these repeat ops are not yet supported in tt-mlir, created below issues.

1. https://github.com/tenstorrent/tt-mlir/issues/1369
2. https://github.com/tenstorrent/tt-mlir/issues/1370

### Logs
[test_mistral_fail.log](https://github.com/user-attachments/files/17867719/test_mistral_fail.log)

nvukobratTT commented on forge/forge/op/tm.py:
```diff
@@ -452,6 +452,29 @@ def Repeat(name: str, operandA: Tensor, factors: List[int]) -> Tensor:
 
 
+def RepeatInterleave(name: str, operandA: Tensor, repeats: int, dim: int) -> Tensor:
+    """
```
Let's add more description about this op.

ashokkumarkannan1 replied:
Done

nvukobratTT commented on forge/forge/tvm_to_python.py:
```diff
@@ -1739,6 +1749,7 @@ def populate_requantize_args(graph, nid, compiler_cfg):
     "take": "take",
     "tanh": "tanh",
     "tile": "repeat",
+    "repeat": "repeat_interleave",
```
Shouldn't repeat be targeted to the pure repeat op? Not repeat interlived?

ashokkumarkannan1 replied:
Actually 
1. `torch.repeat` op is equivalent to `tvm.relay.tile` and `ttnn.repeat` op
2. `torch.repeat_interleave` op is equivalent to `tvm.relay.repeat` and `ttnn.repeat_interleave` op

Due to this tvm's tile ops is mapped to forge's repeat which will be mapped to ttnn's repeat op and tvm's repeat is mapped to forge's repeat_interleave which will be mapped to ttnn.repeat_interleave op.

nvukobratTT commented on forge/test/mlir/test_ops.py:
```diff
@@ -1909,3 +1909,85 @@ def forward(self, a, b):
+
+
+@pytest.mark.xfail(
+    reason="RuntimeError: Found Unsupported operations while lowering from TTForge to TTIR in forward graph - repeat_interleave"
```
Shouldn't this one fail with ttnn.repeat op as required?

ashokkumarkannan1 replied:
There is post optimise decomposition for repeat op where it is decomposed to repeat_interleave op. We can remove that too as we asked the support for `ttnn.repeat` op support in MLIR

nvukobratTT replied:
If I understood correctly:
1. torch.repeat == numpy.tile/tvm.tile == forge.repeat == ttnn.repeat?
2. torch.repeat_interleave == numpy.repeat/tvm.repeat == forge.repeat_interleave == ttnn.repeat_interleave?

If yes, let's outline this clearly and provide more detailed explanations in op description. Have in mind that this can be as a Note section of op documentation :))

ashokkumarkannan1 replied:
Yes @nvukobratTT . I will add this note in op description as you suggested

nvukobratTT commented on forge/test/mlir/test_ops.py:
```diff
@@ -1909,3 +1909,85 @@ def forward(self, a, b):
+            self.expand_shape = expand_shape
+
+        def forward(self, x):
+            x = x + x
```
Why do we need this one?

ashokkumarkannan1 replied:
Previously I was hit with `"There should only be one cpu pre function"` but in latest main I'm not getting this. Will remove this and update the PR

ashokkumarkannan1 replied:
Updated

nvukobratTT commented on forge/test/mlir/test_ops.py:
```diff
@@ -1904,3 +1904,82 @@ def forward(self, a, b):
+
+
+@pytest.mark.xfail(
+    reason="RuntimeError: Found Unsupported operations while lowering from TTForge to TTIR in forward graph - repeat"
```
Make sure to open issues on MLIR side attached to the Bringup milestone with the requirement of adding these ops. Something like this:
- https://github.com/tenstorrent/tt-mlir/issues/1063

nvukobratTT commented:
2 minor comments:
- Op docs updates
- Reminder to open up MLIR issues

Once those are corrected, feel free to merge these changes :))

ashokkumarkannan1 commented:
> 2 minor comments:
> 
> * Op docs updates
> * Reminder to open up MLIR issues
> 
> Once those are corrected, feel free to merge these changes :))

Addressed all your comments and Thanks for your comments :)
